doc/specs/#605 - Search/spec.md

```diff
@@ -0,0 +1,109 @@
+8. To avoid searchBox from blocking the terminal window, the search box could be moved to the bottom right by clicking a button. 
+
+Conhost already has a module for search. It realizes case sensitive or insensitive exact text match search, and it provides methods to select the found word. However, we want to make search as a shared component between Terminal and Console host. Now search module is part of Conhost, and its dependencies include BufferOut and some other types in ConHost such as SCREEN_INFORMATION. In order to make Search a shared component, we need to remove its dependency on ConHost types. BufferOut is already a shared component, but we need to make sure there is no other Conhost dependency.
+
+We will create a SearchBoxControl Xaml UserControl element. When a search process begins, a SearchBoxControl object will be created and attach to TermControl root grid. In other words, one SearchBox is added for each TermControl. The reasons for this design is:
+
+1. Each TermControl corresponds to an iterative Terminal Window and has a individual text buffer. In phrase 1 we are going to search witin the current terminal text buffer. 
+2. If we put the search box under TerminalApp, then the search can only happen on the current focused Terminal. 
+3. If the community does not like the current design, we can lift SearchBox to a higher level. 
+
+### Search process
+1. The user press Ctrl+F (or user's custom key binding) to open the search box. Focus will move to the TextBox. 
+2. Search is performed on a XAML TextBox. Once the user presses Enter, we start to search from the last line of the current viewport or the current selection, and try to find the exact text in the text buffer. The nearest searched one will be selected. Then the search start point will be set to the selected text. The next time "Find" button is clicked, the search will start before or after the previous searched text.
+3. The user can choose to search up or down by choosing up arrow or down arrow buttons. The chosen button will be styled to indicate it is clicked. If the user does not click the arrows buttons, the default direction is up. 
+4. The user can choose to do case sensitive or insensitive match by checking a check box. 
+5. If the search box is focused, the user can navigte all the elements on the search box using tab. When selected, press Enter equals to click. 
+6. If the user click the "X" button or press Esc, the search stopped and the search box disappears and focus will move back to Terminal. In phrase one we do not store any state. 
+7. Once the search box is closed (exiting search mode), the selection will still be there. This coincides with the current VS Code and cmd experience. To get rid of the selection, the user can just click other area of the window.
+

```

Comment:
What happens if the user clicks on the terminal content itself, drawing focus from the find dialog? Does the dialog stay open? Does it dismiss the current selection, but keep the dialog open? Can the user scroll the terminal content, or otherwise input text into the terminal while the dialog is open?

Similarly, what happens if the user switches tabs while the find dialog is open? Does re-focusing the tab move focus to the find dialog of the focused `TermControl`, or does it move focus to the terminal itself?

Reply:
1. What happens if the user clicks on the terminal content itself, drawing focus from the find dialog? Does the dialog stay open? Does it dismiss the current selection, but keep the dialog open? 

Yes, the search dialog stays open. The current selection is dismissed because of the click. 

2. Can the user scroll the terminal content, or otherwise input text into the terminal while the dialog is open?

Yes, the user can scroll or input text into the terminal content even if the search dialog is open. The user can have full interaction with the terminal. 

3. what happens if the user switches tabs while the find dialog is open? Does re-focusing the tab move focus to the find dialog of the focused TermControl, or does it move focus to the terminal itself?

Neither. It will not focus on the TermControl. I will fix this.

Reply:
Update: I synced and with the new changes, switching between tabs is working very well. Once a tab is re-focused with the search dialog open, focus will still be moved to Terminal.

Reply:
Added.